examples/stabilizer_code.py

```diff
@@ -0,0 +1,340 @@
+    The matrix is converted into another matrix with as many rows, but this time the vectors
+    contain the letters I, X, Y, and Z representing Pauli operators.
+    """
+    out = []
+    n = mat.shape[1] // 2
+    for i in range(mat.shape[0]):
+        ps = ''
+        for j in range(n):
+            if mat[i, j] == 0 and mat[i, j + n] == 0:
+                ps += 'I'
+            elif mat[i, j] == 1 and mat[i, j + n] == 0:
+                ps += 'X'
+            elif mat[i, j] == 0 and mat[i, j + n] == 1:
+                ps += 'Z'
+            else:
+                ps += 'Y'
+        out.append(ps)
+    return out
+
+
+# It was considered to use scipy.linalg.lu but it seems to be only for real numbers and does
+# not allow to restrict only on a section of the matrix.
+def _GaussianElimination(
+    M: np.ndarray, min_row: int, max_row: int, min_col: int, max_col: int
+) -> int:
+    """
+    Performs a Gaussian elemination of the input matrix and transforms it into its reduced row
+    echelon form. The elimination is done only on a sub-section of the matrix (specified) by
+    ranges of rows and columns. The matrix elements are integers {0, 1} interpreted as elements
+    of GF(2).
+
+    In short, this is the implementation of section 4.1 of the thesis.
+
+    Args:
+        M: The input/output matrix
+        min_row: The minimum row (inclusive) where the perform the elimination.
+        max_row: The maximum row (exclusive) where the perform the elimination.
+        min_col: The minimum column (inclusive) where the perform the elimination.
+        max_col: The maximum column (exclusive) where the perform the elimination.
+
+    Returns:
+        The rank of the matrix.
+    """
+    assert M.shape[1] % 2 == 0
+    n = M.shape[1] // 2
+
+    max_rank = min(max_row - min_row, max_col - min_col)
+
+    rank = 0
+    for r in range(max_rank):
+        i = min_row + r
+        j = min_col + r
+        pivot_rows, pivot_cols = np.nonzero(M[i:max_row, j:max_col])
+
+        if pivot_rows.size == 0:
+            break
+
+        pi = pivot_rows[0]
+        pj = pivot_cols[0]
+
+        # Swap the rows:
+        M[[i, i + pi]] = M[[i + pi, i]]
+
+        # Swap the columns:
+        M[:, [(j + pj), j]] = M[:, [j, (j + pj)]]
+
+        # Since the columns in the left and right half of the matrix represent the same qubit, we
+        # also need to swap the corresponding column in the other half.
+        j_other_half = (j + n) % (2 * n)
+        M[:, [(j_other_half + pj), j_other_half]] = M[:, [j_other_half, (j_other_half + pj)]]
+
+        # Do the elimination.
+        for k in range(i + 1, max_row):
+            if M[k, j] == 1:
+                M[k, :] = np.mod(M[i, :] + M[k, :], 2)
+
+        rank += 1
+
+    # Backward replacing to get identity
+    for r in reversed(range(rank)):
+        i = min_row + r
+        j = min_col + r
+
+        # Do the elimination.
+        for k in reversed(range(min_row, i)):
```

Comment:
`reversed` must drain all elements from `range` and store them before it can begin returning them in the reversed order. You can avoid this by using `for k in range(i - 1, min_row - 1, -1)` (this will include `min_row` and exclude `i` the same as `reversed(range(min_row, i))`).

Reply:
Thanks, I didn't know about the behavior of reversed. Done.